### PR TITLE
fix: additional newline added to docstring when the previous line length is less than the line length limit minus 1

### PR DESCRIPTION
**Pull request by veryslowcode**
### Description

     This resolves the issue mentioned in issue #4180, where a newline will
     be added to a docstring comment (that is approaching the line limit),
     even if the docstring contains a newline before the closing quotes.

### Checklist - did you ...

- [✓] Add an entry in `CHANGES.md` if necessary?
- [✓] Add / update tests if necessary?
- [✓] Add new / update outdated documentation?

### CODE REVIEW

File: src/black/linegen.py [hunk 1]
```diff
@@ -477,13 +477,15 @@ def visit_STRING(self, leaf: Leaf) -> Iterator[Line]:
...
-                # closing quotes
+                # the maximum line length, and the closing quote is not
+                # prefixed by a newline then put a line break before
+                # the closing quotes
                 if (
                     len(lines) > 1
                     and last_line_length + quote_len > self.mode.line_length
                     and len(indent) + quote_len <= self.mode.line_length
                     and not has_trailing_backslash
+                    and not leaf.value[-4] == "\n"
```
Comment (JelleZijlstra) [inline src/black/linegen.py:488]: This will need to be gated to the preview style (see #4164 for a recent example).

Comment (veryslowcode) [reply src/black/linegen.py]: Thank you for the example. With the most recent push, I believe this should be gated now.

**Comment by veryslowcode** [comment]
I see that diff-shades detected changes, but they appear to be the intended changes based on the original issue. If there is anything remaining that needs to be done/changed, please let me know and I am happy to do so.

File: src/black/mode.py [hunk 2]
```diff
@@ -177,6 +177,7 @@ class Preview(Enum):
     wrap_long_dict_values_in_parens = auto()
     multiline_string_handling = auto()
     typed_params_trailing_comma = auto()
+    docstring_check_for_newline = auto()
```
Comment (JelleZijlstra) [inline src/black/mode.py:180]: Could you document this in future-style.md? (I have a PR in #4187 that will enforce this.)

File: src/black/linegen.py [hunk 3]
```diff
@@ -477,15 +477,22 @@ def visit_STRING(self, leaf: Leaf) -> Iterator[Line]:
...
                 if (
                     len(lines) > 1
                     and last_line_length + quote_len > self.mode.line_length
                     and len(indent) + quote_len <= self.mode.line_length
                     and not has_trailing_backslash
                 ):
-                    leaf.value = prefix + quote + docstring + "\n" + indent + quote
+                    if (
+                        Preview.docstring_check_for_newline in self.mode
+                        and leaf.value[-4] == "\n"
```
Comment (JelleZijlstra) [inline src/black/linegen.py:491]:
```suggestion
                        and leaf.value[-1 - quote_len] == "\n"
```

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, this is close to being ready.

File: src/black/mode.py (hunk 2 above)
Comment (veryslowcode) [reply src/black/mode.py:180]: Yep, it is in there now. I do apologize for missing that.

**Pull request merged**